packages/i18n/src/index.tsx

```diff
@@ -157,10 +158,98 @@ export const createI18nContext = (
      */
     dict: (lang: string) => deepReadObject(dict, lang)
   };
-  return [translate, actions];
+  return [translate, actions as any];
 };
 
 type I18nContextInterface = ReturnType<typeof createI18nContext>;
 
 export const I18nContext = createContext<I18nContextInterface>({} as I18nContextInterface);
 export const useI18n = () => useContext(I18nContext);
+
+// -------------------------- Chained I18n -------------------
+type I18nFormatArgs = Record<string, string | number>;
+
+export type I18nPath<T> = {
+  [K in keyof T]: T[K] extends Record<string, unknown>
+    ? I18nPath<T[K]>
+    : T[K] extends (...args: any) => string
+    ? (...args: Parameters<T[K]>) => string
+    : T[K] extends string
+    ? (args?: I18nFormatArgs) => string
+    : never;
+};
+
+const buildI18nChain = <T,>(obj: T): I18nPath<T> => {
+  const keys = Object.keys(obj as any) as (keyof T)[];
+  const paths = keys.reduce((acc, key) => {
+    const value = obj[key];
+    if (typeof value === "object") {
+      return {
+        ...acc,
+        [key]: buildI18nChain(value)
+      } as any;
+    } else if (typeof value === "function") {
+      return {
+        ...acc,
+        [key]: obj[key]
+      };
+    }
+    if (typeof value === "string") {
+      return {
+        ...acc,
+        [key]: (args: I18nFormatArgs) => {
+          return value.replace(/{{(.*?)}}/g, (_, key) => deepReadObject(args, key, ""));
+        }
+      };
+    } else {
+      throw new TypeError("Unsupported data format on the keys");
+    }
+  }, {} as Partial<I18nPath<T>>);
+
+  return paths as I18nPath<T>;
+};
+
+export const makeChainedI18nContext = <T extends object, K extends keyof T>(props: {
+  dictionaries: T;
+  locale: keyof T;
+}) => {
+  const [store, setStore] = createStore<{ dictionaries: T; locale: keyof T }>(props);
+
+  const utils = {
+    locale(locale?: K): K {
+      if (locale) {
+        setStore("locale", locale);
+        return locale;
+      }
+      return store.locale as K;
+    },
+    changeLocale(locale: K) {
+      setStore("locale", locale);
+    },
+    getDictionary(locale?: K): T[K] {
+      if (locale) return store.dictionaries[locale];
+      return store.dictionaries[store.locale as K];
+    }
+  };
+
+  const chainedI18n = createMemo(() => buildI18nChain(store.dictionaries[store.locale]));
+
+  const translate = () => {
+    return chainedI18n();
+  };
+
+  const context = createContext<[typeof translate, typeof utils] | null>(null);
```

Comment:
Maybe you could add a context default that allows a global i18n context, so the provider is optional?

Reply:
Yes. I can add an option to have the context be passed in by default to make it a globally scoped one. Keeping it null by default allows the user to scope things to certain areas and avoid a global dictionary. 

I'll also remove the changeLocale method since it's not needed.

Reply:
Added the option in, removed `changeLocale`, and fixed an issue with the build after I converted it to be a `.tsx` file.